### ISSUE TITLE
issue about gmm_loss

### ISSUE THREAD

**Issue by UPUPGOO** [ctallec/world-models#35]
In the file models/mdrnn.py, why minus max_log_probs first (line 38) and then add max_log_probs.squeeze() (line 43) ?

    batch = batch.unsqueeze(-2)
    normal_dist = Normal(mus, sigmas)
    g_log_probs = normal_dist.log_prob(batch)
    g_log_probs = logpi + torch.sum(g_log_probs, dim=-1)
    max_log_probs = torch.max(g_log_probs, dim=-1, keepdim=True)[0]
    g_log_probs = g_log_probs - max_log_probs
    g_probs = torch.exp(g_log_probs)
    probs = torch.sum(g_probs, dim=-1)

    log_prob = max_log_probs.squeeze() + torch.log(probs)
    if reduce:
        return - torch.mean(log_prob)
    return - log_prob

Is there a corresponding mathematical formula？

**Comment by UPUPGOO** [comment]
The two are equivalent.

**Comment by pdeubel** [comment]
I also came across this and found it odd to have seemingly no effect. But I think this is done for numerical stability and also known as the "Log-Sum-Exp" Trick, nicely explained here: https://gregorygundersen.com/blog/2020/02/09/log-sum-exp/

I quickly tested it and made a variant not using this trick and for one value I got `-inf` while with the trick I got a valid value
